File: src/math/curve.py
import bisect

from .matrix import matrix_from_quat, matrix_identity
from .quaternion import (quat_from_angle_axis, quat_from_direction, quat_from_directions, quat_from_forward_and_up,
                         quat_from_matrix_largest_component, quat_identity, quat_interpolate, quat_inverse,
                         quat_multiply, quat_rotate, quat_squad)
from .scalar import clamp, epsilon, mix
from .spline import catmull_rom, catmull_rom_length
from .vector import vec3, vec3_add, vec3_distance, vec3_normalize, vec3_subtract, vec4


class Curve:
    class Node:
        __slots__ = ("position", "orientation", "arc_length", "segment_index")

        def __init__(self, position=None, segment_index=0):
            self.position = vec3(0.0, 0.0, 0.0) if position is None else position
            self.orientation = quat_identity
            self.arc_length = 0.0
            self.segment_index = segment_index

    class Span:
        __slots__ = ("from_index", "to_index", "t")

        def __init__(self, from_index=0, to_index=0, t=0.0):
            self.from_index = from_index
            self.to_index = to_index
            self.t = t

    def __init__(self):
        self.node_list = []
        self.segment_boundaries = [0.0]
        self.closed = False

    def insert_node(self, position, segment_index):
        if self.node_list and vec3_distance(self.node_list[-1].position, position) < epsilon:
            return None
        self.node_list.append(Curve.Node(position, segment_index))
        segment_end = segment_index + 1
        if len(self.segment_boundaries) <= segment_end:
            self.segment_boundaries += [0.0] * (segment_end + 1 - len(self.segment_boundaries))

        newest = len(self.node_list) - 1
        if newest == 0:
            return newest
        previous = newest - 1
        if previous > 0:
            self.update_span(previous)
        self.node_list[newest].arc_length = (self.node_list[previous].arc_length
                                             + vec3_distance(self.node_list[previous].position, position))
        self.segment_boundaries[segment_end] = self.node_list[newest].arc_length
        return newest

    def insert_position(self, position, segment_index=0):
        newest = self.insert_node(position, segment_index)
        if newest is None or newest == 0:
            return
        previous = newest - 1
        tangent_at_previous = self.tangent_at(previous)
        if previous == 0:
            self.node_list[0].orientation = quat_from_direction(tangent_at_previous)
        else:
            self.node_list[previous].orientation = quat_multiply(
                quat_from_directions(self.tangent_at(previous - 1), tangent_at_previous),
                self.node_list[previous - 1].orientation)
        self.node_list[newest].orientation = quat_multiply(
            quat_from_directions(tangent_at_previous, self.tangent_at(newest)), self.node_list[previous].orientation)

    def insert_matrix(self, matrix, segment_index=0):
        newest = self.insert_node(vec3(matrix[3][0], matrix[3][1], matrix[3][2]), segment_index)
        if newest is None:
            return
        self.node_list[newest].orientation = quat_from_matrix_largest_component(matrix)

    def close(self):
        self.closed = True
        if len(self.node_list) < 3:
            return
        last = len(self.node_list) - 1
        first_span_before = self.node_list[1].arc_length
        self.update_span(1)
        closing_shift = self.node_list[1].arc_length - first_span_before
        for index in range(2, last + 1):
            self.node_list[index].arc_length += closing_shift
            self.segment_boundaries[self.node_list[index].segment_index + 1] = self.node_list[index].arc_length
        self.update_span(last)

    def close_orientations(self):
        if len(self.node_list) < 3:
            return
        last = len(self.node_list) - 1
        second_node_before_closing = self.node_list[1].orientation
        self.node_list[0].orientation = quat_from_direction(self.tangent_at(0))
        self.parallel_transport_node(1)
        closing_twist = quat_multiply(quat_inverse(second_node_before_closing), self.node_list[1].orientation)
        for index in range(2, last):
            self.node_list[index].orientation = quat_multiply(self.node_list[index].orientation, closing_twist)
        self.parallel_transport_node(last)

    def apply_roll(self, roll_at):
        forward_axis = vec3(0.0, 0.0, 1.0)
        for node in self.node_list:
            node.orientation = quat_multiply(node.orientation,
                                                   quat_from_angle_axis(-roll_at(node.arc_length), forward_axis))

    def with_offset(self, offset):
        offset_curve = self.copy()
        offset_in_frame = vec3(offset[0], offset[1], 0.0)
        for index in range(len(self.node_list)):
            offset_curve.node_list[index].position = vec3_add(
                self.node_list[index].position,
                quat_rotate(self.node_list[index].orientation, offset_in_frame))
        for index in range(1, len(self.node_list)):
            offset_curve.update_span(index)

        def orient_along_offset_path(index):
            up = quat_rotate(self.node_list[index].orientation, vec3(0.0, 1.0, 0.0))
            offset_curve.node_list[index].orientation = quat_from_forward_and_up(
                offset_curve.tangent_at(index), up)

        first_movable = 0 if self.closed else 1
        last_movable = len(self.node_list) if self.closed else len(self.node_list) - 1
        for index in range(first_movable, last_movable):
            orient_along_offset_path(index)
        return offset_curve

    def matrix_at(self, arc_length):
        if not self.node_list:
            return matrix_identity
        if len(self.node_list) == 1:
            return self.matrix_at_node(0)
        span = self.span_at(arc_length)
        rows = matrix_from_quat(self.orientation_at_span(span))
        position = catmull_rom(self.node_at(span.from_index - 1).position, self.node_list[span.from_index].position,
                               self.node_list[span.to_index].position, self.node_at(span.to_index + 1).position,
                               span.t)
        return (rows[0], rows[1], rows[2], vec4(position[0], position[1], position[2], 1.0))

    def matrix_at_node(self, node_index):
        node = self.node_at(node_index)
        rows = matrix_from_quat(node.orientation)
        return (rows[0], rows[1], rows[2], vec4(node.position[0], node.position[1], node.position[2], 1.0))

    def orientation_at(self, arc_length):
        if len(self.node_list) < 2:
            return quat_identity if not self.node_list else self.node_list[0].orientation
        return self.orientation_at_span(self.span_at(arc_length))

    def orientation_between_nodes_at(self, arc_length):
        if len(self.node_list) < 2:
            return self.orientation_at(arc_length)
        span = self.span_at(arc_length)
        return quat_interpolate(self.node_list[span.from_index].orientation, self.node_list[span.to_index].orientation, span.t)

    def arc_length_at(self, segment_coordinate):
        segment_count = self.number_of_segments()
        if segment_count <= 0:
            return 0.0
        coordinate = clamp(segment_coordinate, 0.0, float(segment_count))
        segment = min(int(coordinate), segment_count - 1)
        segment_start = self.segment_boundaries[segment]
        segment_end = self.segment_boundaries[segment + 1]
        return mix(segment_start, segment_end, coordinate - segment)

    def segment_coordinate_at(self, arc_length):
        segment_count = self.number_of_segments()
        if segment_count <= 0:
            return 0.0
        first_end_beyond = bisect.bisect_right(self.segment_boundaries, arc_length, lo=1)
        segment = clamp(first_end_beyond - 1, 0, segment_count - 1)
        segment_start = self.segment_boundaries[segment]
        segment_length = self.segment_boundaries[segment + 1] - segment_start
        fraction = (arc_length - segment_start) / segment_length if segment_length > epsilon else 0.0
        return segment + clamp(fraction, 0.0, 1.0)

    def nodes(self):
        return self.node_list

    def arc_length(self):
        return 0.0 if not self.node_list else self.node_list[-1].arc_length

    def number_of_segments(self):
        return len(self.segment_boundaries) - 1

    def copy(self):
        copied = Curve()
        copied.closed = self.closed
        copied.segment_boundaries = list(self.segment_boundaries)
        for node in self.node_list:
            copied_node = Curve.Node(node.position, node.segment_index)
            copied_node.orientation = node.orientation
            copied_node.arc_length = node.arc_length
            copied.node_list.append(copied_node)
        return copied

    def span_at(self, arc_length):
        first_node_beyond = bisect.bisect_right(self.node_list, arc_length, key=lambda node: node.arc_length)
        to_index = clamp(first_node_beyond, 1, len(self.node_list) - 1)
        span_length = self.node_list[to_index].arc_length - self.node_list[to_index - 1].arc_length
        t = (clamp((arc_length - self.node_list[to_index - 1].arc_length) / span_length, 0.0, 1.0)
             if span_length > epsilon else 0.0)
        return Curve.Span(to_index - 1, to_index, t)

    def orientation_at_span(self, span):
        return quat_squad(self.node_at(span.from_index - 1).orientation, self.node_list[span.from_index].orientation,
                     self.node_list[span.to_index].orientation, self.node_at(span.to_index + 1).orientation, span.t)

    def node_at(self, index):
        count = len(self.node_list)
        if self.closed and count > 2:
            ring_length = count - 1
            return self.node_list[(index % ring_length + ring_length) % ring_length]
        return self.node_list[clamp(index, 0, count - 1)]

    def tangent_at(self, index):
        return vec3_normalize(vec3_subtract(self.node_at(index + 1).position, self.node_at(index - 1).position))

    def parallel_transport_node(self, index):
        self.node_list[index].orientation = quat_multiply(
            quat_from_directions(self.tangent_at(index - 1), self.tangent_at(index)), self.node_list[index - 1].orientation)

    def update_span(self, index):
        node = self.node_list[index]
        node.arc_length = self.node_list[index - 1].arc_length + catmull_rom_length(
            self.node_at(index - 2).position, self.node_list[index - 1].position, node.position,
            self.node_at(index + 1).position)
        self.segment_boundaries[node.segment_index + 1] = node.arc_length

File: src/curve.cpp
#include "curve.hpp"
#include "math.hpp"
#include "spline.hpp"

#include <glm/common.hpp>
#include <glm/geometric.hpp>
#include <glm/gtx/quaternion.hpp>

#include <algorithm>
#include <cmath>

namespace nolimits2track {

Curve::Curve(std::size_t node_count) {
    node_list.reserve(node_count);
}

void Curve::insert_position(const glm::dvec3& position, int segment_index) {
    if (!node_list.empty() && glm::distance(node_list.back().position, position) < epsilon) {
        return;
    }
    node_list.push_back({.position = position, .segment_index = segment_index});
    const std::size_t segment_end = static_cast<std::size_t>(segment_index) + 1;
    if (segment_boundaries.size() <= segment_end) {
        segment_boundaries.resize(segment_end + 1, 0.0);
    }

    const std::size_t newest = node_list.size() - 1;
    if (newest == 0) {
        return;
    }
    const std::size_t previous = newest - 1;
    if (previous > 0) {
        update_span(previous);
    }
    node_list[newest].arc_length = node_list[previous].arc_length + glm::distance(node_list[previous].position, position);
    segment_boundaries[segment_end] = node_list[newest].arc_length;

    const glm::dvec3 tangent_at_previous = tangent_at(previous);
    if (previous == 0) {
        node_list[0].orientation = orientation_from_direction(tangent_at_previous);
    } else {
        node_list[previous].orientation =
            glm::rotation(tangent_at(previous - 1), tangent_at_previous) * node_list[previous - 1].orientation;
    }
    node_list[newest].orientation = glm::rotation(tangent_at_previous, tangent_at(newest)) * node_list[previous].orientation;
}

void Curve::close() {
    closed = true;
    if (node_list.size() < 3) {
        return;
    }
    const std::size_t last = node_list.size() - 1;
    const double first_span_before = node_list[1].arc_length;
    update_span(1);
    const double seam_shift = node_list[1].arc_length - first_span_before;
    for (std::size_t index = 2; index <= last; ++index) {
        node_list[index].arc_length += seam_shift;
        segment_boundaries[static_cast<std::size_t>(node_list[index].segment_index) + 1] = node_list[index].arc_length;
    }
    update_span(last);

    const glm::dquat second_node_before_closing = node_list[1].orientation;
    node_list[0].orientation = orientation_from_direction(tangent_at(0));
    parallel_transport_node(1);
    const glm::dquat closing_twist = glm::inverse(second_node_before_closing) * node_list[1].orientation;
    for (std::size_t index = 2; index < last; ++index) {
        node_list[index].orientation = node_list[index].orientation * closing_twist;
    }
    parallel_transport_node(last);
}

void Curve::apply_roll(const std::function<double(double)>& roll_at) {
    const glm::dvec3 forward_axis = {0.0, 0.0, 1.0};
    for (Node& node : node_list) {
        node.orientation = node.orientation * glm::angleAxis(-roll_at(node.arc_length), forward_axis);
    }
}

Curve Curve::with_offset(const glm::dvec2& offset) const {
    Curve offset_curve = *this;
    for (std::size_t index = 0; index < node_list.size(); ++index) {
        offset_curve.node_list[index].position =
            node_list[index].position + node_list[index].orientation * glm::dvec3(offset, 0.0);
    }
    for (std::size_t index = 1; index < node_list.size(); ++index) {
        offset_curve.update_span(index);
    }

    const auto orient_along_offset_path = [&](std::size_t index) {
        const glm::dvec3 up = node_list[index].orientation * glm::dvec3(0.0, 1.0, 0.0);
        offset_curve.node_list[index].orientation = orientation_from_forward_and_up(offset_curve.tangent_at(index), up);
    };
    const std::size_t first_movable = closed ? 0 : 1;
    const std::size_t last_movable = closed ? node_list.size() : node_list.size() - 1;
    for (std::size_t index = first_movable; index < last_movable; ++index) {
        orient_along_offset_path(index);
    }
    return offset_curve;
}

glm::dmat4 Curve::matrix_at(double arc_length) const {
    if (node_list.empty()) {
        return glm::dmat4(1.0);
    }
    if (node_list.size() == 1) {
        return matrix_at(std::size_t{0});
    }
    const Span span = span_at(arc_length);
    glm::dmat4 frame = glm::mat4_cast(orientation_at(span));
    frame[3] = glm::dvec4(catmull_rom(node_at(span.from - 1).position, node_list[span.from].position,
                                     node_list[span.to].position, node_at(span.to + 1).position, span.t),
                          1.0);
    return frame;
}

glm::dmat4 Curve::matrix_at(std::size_t node_index) const {
    const Node& node = node_at(node_index);
    glm::dmat4 frame = glm::mat4_cast(node.orientation);
    frame[3] = glm::dvec4(node.position, 1.0);
    return frame;
}

glm::dquat Curve::orientation_at(double arc_length) const {
    if (node_list.size() < 2) {
        return node_list.empty() ? glm::dquat(1.0, 0.0, 0.0, 0.0) : node_list[0].orientation;
    }
    return orientation_at(span_at(arc_length));
}

glm::dquat Curve::orientation_between_nodes_at(double arc_length) const {
    if (node_list.size() < 2) {
        return orientation_at(arc_length);
    }
    const Span span = span_at(arc_length);
    return slerp(node_list[span.from].orientation, node_list[span.to].orientation, span.t);
}

double Curve::arc_length_at(double segment_coordinate) const {
    const int segment_count = number_of_segments();
    if (segment_count <= 0) {
        return 0.0;
    }
    const double coordinate = glm::clamp(segment_coordinate, 0.0, static_cast<double>(segment_count));
    const int segment = std::min(static_cast<int>(std::floor(coordinate)), segment_count - 1);
    const double segment_start = segment_boundaries[static_cast<std::size_t>(segment)];
    const double segment_end = segment_boundaries[static_cast<std::size_t>(segment) + 1];
    return glm::mix(segment_start, segment_end, coordinate - segment);
}

double Curve::segment_coordinate_at(double arc_length) const {
    const int segment_count = number_of_segments();
    if (segment_count <= 0) {
        return 0.0;
    }
    const auto first_end_beyond = std::upper_bound(segment_boundaries.begin() + 1, segment_boundaries.end(), arc_length);
    const int segment =
        glm::clamp(static_cast<int>(first_end_beyond - segment_boundaries.begin()) - 1, 0, segment_count - 1);
    const double segment_start = segment_boundaries[static_cast<std::size_t>(segment)];
    const double segment_length = segment_boundaries[static_cast<std::size_t>(segment) + 1] - segment_start;
    const double fraction = segment_length > epsilon ? (arc_length - segment_start) / segment_length : 0.0;
    return segment + glm::clamp(fraction, 0.0, 1.0);
}

Curve::Span Curve::span_at(double arc_length) const {
    const auto first_node_beyond = std::upper_bound(
        node_list.begin(), node_list.end(), arc_length,
        [](double value, const Node& node) { return value < node.arc_length; });
    const std::size_t to = glm::clamp<std::size_t>(
        static_cast<std::size_t>(first_node_beyond - node_list.begin()), 1, node_list.size() - 1);
    const double span_length = node_list[to].arc_length - node_list[to - 1].arc_length;
    const double t =
        span_length > epsilon ? glm::clamp((arc_length - node_list[to - 1].arc_length) / span_length, 0.0, 1.0) : 0.0;
    return {.from = to - 1, .to = to, .t = t};
}

glm::dquat Curve::orientation_at(const Span& span) const {
    return squad(node_at(span.from - 1).orientation, node_list[span.from].orientation, node_list[span.to].orientation,
                 node_at(span.to + 1).orientation, span.t);
}

const Curve::Node& Curve::node_at(std::size_t index) const {
    const std::ptrdiff_t at = static_cast<std::ptrdiff_t>(index);
    const std::ptrdiff_t count = static_cast<std::ptrdiff_t>(node_list.size());
    if (closed && count > 2) {
        const std::ptrdiff_t ring_length = count - 1;
        return node_list[static_cast<std::size_t>((at % ring_length + ring_length) % ring_length)];
    }
    return node_list[static_cast<std::size_t>(std::clamp<std::ptrdiff_t>(at, 0, count - 1))];
}

glm::dvec3 Curve::tangent_at(std::size_t index) const {
    return glm::normalize(node_at(index + 1).position - node_at(index - 1).position);
}

void Curve::parallel_transport_node(std::size_t index) {
    node_list[index].orientation = glm::rotation(tangent_at(index - 1), tangent_at(index)) * node_list[index - 1].orientation;
}

void Curve::update_span(std::size_t index) {
    Node& node = node_list[index];
    node.arc_length = node_list[index - 1].arc_length
                     + catmull_rom_length(node_at(index - 2).position, node_list[index - 1].position, node.position,
                                        node_at(index + 1).position);
    segment_boundaries[static_cast<std::size_t>(node.segment_index) + 1] = node.arc_length;
}

const std::vector<Curve::Node>& Curve::nodes() const {
    return node_list;
}

double Curve::arc_length() const {
    return node_list.empty() ? 0.0 : node_list.back().arc_length;
}

int Curve::number_of_segments() const {
    return static_cast<int>(segment_boundaries.size()) - 1;
}

}
